pkg/auth/auth.go

```diff
@@ -1,66 +1,137 @@
+	Repository   string
 
-	auth := &Authorization{Endpoints: []Endpoint{}}
-	for _, r := range c.Repositories {
-		git, err := regexp.Compile(`/` + c.Organization + `/` + r.Project + `/_git/` + r.Name + `(/.*)?\b`)
-		if err != nil {
-			return nil, err
-		}
+	regexes []*regexp.Regexp
+}
 
-		api, err := regexp.Compile(`/` + c.Organization + `/` + r.Project + `/_apis/git/repositories/` + r.Name + `(/.*)?\b`)
+// NewAuthorization cretes a new authorization from a give configuration.
+func NewAuthorization(cfg config.Configuration) (Authorization, error) {
+	authz := Authorization{endpoints: map[string]Endpoint{}}
+	for _, o := range cfg.Organizations {
+		baseApi, err := regexp.Compile(fmt.Sprintf(`/%s/_apis\b`, o.Name))
 		if err != nil {
-			return nil, err
+			return Authorization{}, fmt.Errorf("invalid base api regex: %v", err)
 		}
 
-		endpoint := Endpoint{
-			Token:   r.Token,
-			Regexes: []*regexp.Regexp{baseApi, git, api},
+		for _, r := range o.Repositories {
+			git, err := regexp.Compile(fmt.Sprintf(`/%s/%s/_git/%s(/.*)?\b`, o.Name, r.Project, r.Name))
+			if err != nil {
+				return Authorization{}, err
+			}
+			api, err := regexp.Compile(fmt.Sprintf(`/%s/%s/_apis/git/repositories/%s(/.*)?\b`, o.Name, r.Project, r.Name))
+			if err != nil {
+				return Authorization{}, err
+			}
+			token, err := randomSecureToken()
+			if err != nil {
+				return Authorization{}, fmt.Errorf("could not generate random token: %v", err)
+			}
+
+			e := Endpoint{
+				Pat:          o.Pat,
+				Domain:       o.Domain,
+				Scheme:       o.Scheme,
+				Token:        token,
+				Namespaces:   r.Namespaces,
+				SecretName:   o.GetSecretName(r),
+				Organization: o.Name,
+				Project:      r.Project,
+				Repository:   r.Name,
+				regexes:      []*regexp.Regexp{baseApi, git, api},
+			}
+			authz.endpoints[token] = e
 		}
-		auth.Endpoints = append(auth.Endpoints, endpoint)
 	}
 
-	return auth, nil
+	return authz, nil
 }
 
-// IsPermitted checks if a specific user is permitted to access a path
-func IsPermitted(a *Authorization, path string, token string) error {
-	for _, e := range a.Endpoints {
-		// Only check regex for matching tokens
-		if !strings.EqualFold(e.Token, token) {
-			continue
+func (a *Authorization) GetEndpoints() map[string]Endpoint {
+	return a.endpoints
+}
+
+// LookupEndpoint returns the endpoint with the matching organization, project and repository.
+// TODO (Philip): Include domain in this lookup
+func (a *Authorization) LookupEndpoint(org, proj, repo string) (Endpoint, error) {
+	for _, v := range a.endpoints {
+		if v.Organization == org && v.Project == proj && v.Repository == repo {
+			return v, nil
 		}
+	}
+	return Endpoint{}, errors.New("endpoint not found")
+}
 
-		// Return of a regex matches the path
-		for _, r := range e.Regexes {
-			if r.MatchString(path) {
-				return nil
-			}
+// PatForToken returns the pat associated with the token
+func (a *Authorization) PatForToken(token string) (string, error) {
```

Comment:
Using different naming convention for the methods. Standardizing on a specific scheme would make it easier to understand.

`GetTokenFromPAT()`